File: two-numbers.php
<?php
/*
* Leetcode: 2. Add Two Numbers LinkedList;
*
*
*/

/**
 * Definition for a singly-linked list.
 */
class ListNode {
    public $val = 0;
    public $next = null;
    function __construct($val = 0, $next = null) {
        $this->val = $val;
        $this->next = $next;
    }
}

class Solution {

    /**
     * @param ListNode $l1
     * @param ListNode $l2
     * @return ListNode
     */
    function addTwoNumbers($l1, $l2) {
        $dummy = new ListNode();
        $current = $dummy;

        $carry = 0;
        while( $l1 || $l2 || $carry ){
            $total = 0;
            if($l1 && $l1->val)
                $total += $l1->val;

            if($l2 && $l2->val)
                $total += $l2->val;

            $total += $carry;

            $carry = intval($total/10);
            $total = $total % 10;
            if( ! $current )
                $current = new ListNode($total);
            else
                $current->next = new ListNode($total);

            $current = $current->next;
            if($l1->next)
                $l1 = $l1->next;
            else
                $l1 = null;

            if($l2->next)
                $l2 = $l2->next;
            else
                $l2 = $l2->next;

        }
        return $dummy->next;
    }

    function print_numbers($nums)
    {
        echo "\n";
        $current = $nums;
        while($current)
        {
            echo $current->val.' ';
            $current = $current->next;
        }
    }
}

$l1 = new ListNode(2);
$l1->next = new ListNode(4);
$l1->next->next = new ListNode(8);

$l2 = new ListNode( 5 );
$l2->next = new ListNode( 6 );
$l2->next->next = new ListNode(4);

$sol = new Solution();
$ans = $sol->addTwoNumbers($l1, $l2);
$sol->print_numbers($l1);
$sol->print_numbers($l2);
$sol->print_numbers($ans);

File: two-numbers.py
# Definition for singly-linked list.
class ListNode(object):
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

class Solution(object):
    def addTwoNumbers(self, l1, l2):
        """
        :type l1: ListNode
        :type l2: ListNode
        :rtype: ListNode
        """
        dummy = ListNode()
        current = dummy

        carry = 0
        while l1 or  l2 or carry:
            val1 = l1.val if l1 else 0
            val2 = l2.val if l2 else 0

            val = val1 + val2 + carry
            carry = val //10
            val = val % 10
            current.next = ListNode(val)

            current = current.next
            l1 = l1.next if l1 else None
            l2 = l2.next if l2 else None

        return dummy.next

    def printList(self, node):
        print('')
        while(node):
            print(node.val, end = ' ')
            node = node.next

l1 = ListNode(2)
l1.next = ListNode(4)
l1.next.next = ListNode(8)

l2 = ListNode(5)
l2.next = ListNode(6)
l2.next.next = ListNode(4)

sol = Solution()
ans = sol.addTwoNumbers(l1, l2)
sol.printList(l1)
sol.printList(l2)
sol.printList(ans)
